Add support to the handler to support fetching the ip address and files

mwringe opened the pull request:
Adds support to the handler to be able to fetch the ip address of a container and the files from within a container.

Support added for docker containers, but if someone knows more about the raw or rkt containers, they should also be updated to add in this support.

Being able to fetch the ip address of container is essential for supporting custom metrics. Without this we need to be able to know the ip address before starting the container. This is especially cumbersome with kuberentes as it requires setting up hostports and manually managing each container that wants to expose metrics. The support added here will work with both plain docker images and kubernetes pods.

Note: this is in piece of the PR sent for https://github.com/google/cadvisor/pull/1328 as it has been deemed more appropriate to split that request into smaller separate ones.

jimmidyson commented:
@k8s-bot Test this

mwringe commented:
Hmm, anyone know what is going on with the e2e tests?

jimmidyson commented:
@k8s-bot ok to test

jimmidyson commented:
@vishh Any idea on the e2e tests?

mwringe commented:
Bump, anything I can do to get this pushed through any faster?

timstclair commented:
From the e2e logs:
```
+ ./build/check_gofmt.sh .
The following files are not properly formatted:
container/docker/handler.go container/docker/handler.go
Build step 'Execute shell' marked build as failure
```

Looks like you need to run gofmt on `container/docker/handler.go` and `container/docker/handler.go`.

timstclair commented on container/mock.go:
```diff
@@ -97,6 +97,14 @@ func (self *MockContainerHandler) Type() ContainerType {
 }
 
+func (self *MockContainerHandler) GetContainerIPAddress() string {
+	return "123.456.789.1011"
```
nit: return a valid IP address

timstclair replied:
The mock should be configurable. Just follow the pattern of the other methods.

timstclair commented on container/docker/handler.go:
```diff
@@ -412,6 +432,21 @@ func (self *dockerContainerHandler) GetContainerLabels() map[string]string {
+func (self *dockerContainerHandler) ReadFile(filepath string) ([]byte, error) {
+	pid := self.pid
+	filePath := path.Join(self.rootFs, "/proc", strconv.Itoa(pid), "/root", filepath)
+	glog.V(3).Infof("Trying path %q", filePath)
```
nit: V(4) for debug info. Add more context (e.g. `"Attempting to read file at %s"`)

timstclair commented on container/docker/handler.go:
```diff
@@ -412,6 +432,21 @@ func (self *dockerContainerHandler) GetContainerLabels() map[string]string {
+	if err == nil {
+		return data, err
+	}
+	return nil, fmt.Errorf("file %q does not exist.", filepath)
```
This is not necessary a not found error.

timstclair commented on container/docker/handler.go:
```diff
@@ -222,6 +226,22 @@ func newDockerContainerHandler(
+	ipAddress := ctnr.NetworkSettings.IPAddress
+	networkNode := string(ctnr.HostConfig.NetworkMode)
+	if ipAddress == "" && strings.HasPrefix(networkNode, "container:") {
+		containerId := networkNode[len("container:"):len(networkNode)]
```
Prefer `strings.TrimPrefix`

timstclair commented on container/docker/handler.go:
```diff
@@ -222,6 +226,22 @@ func newDockerContainerHandler(
+	// If the NetworkMode starts with 'container:' then we need to use the IP address of the container specified.
+	// This happens in cases such as kubernetes where the containers doesn't have an IP address itself and we need to use the pod's address
+	ipAddress := ctnr.NetworkSettings.IPAddress
+	networkNode := string(ctnr.HostConfig.NetworkMode)
```
s/networkNode/networkMode/ ?

timstclair commented on container/container.go:
```diff
@@ -59,6 +59,12 @@ type ContainerHandler interface {
+	GetContainerIPAddress() string
+
+	// Reads a file from within the container
+	ReadFile(path string) ([]byte, error)
```
This looks unused?

mwringe commented:
Ok, I have updated this PR to bring it into line with the latest from master and to take into affect the comments made.

A few changes:
- I removed the ability to read from container directly. Originally I had this in there to allow the ability to pull out a ca cert from a container to verify self signed https endpoints, but I think it makes a lot more sense to just trust any endpoint, regardless if its certificates are necessarily valid or not.

- I also updated the Prometheus and generic collectors to be able to read from a container's ip address directly instead of requiring the ip address to be hardcoded. This will make things a lot more useful when dealing with things such as containers running in Kubernetes

mwringe commented:
Additional commit to make the http client used by the collectors to be configurable. This will allow clients to use certificate based authentication.

timstclair commented on cadvisor.go:
```diff
@@ -186,3 +192,29 @@ func installSignalHandler(containerManager manager.Manager) {
 }
+
+func createCollectorHttpClient(collectorCert, collectorKey string) http.Client {
+	//Enable accessing insecure endpoints. We should be able to access metrics from any endpoint
```
(nit) space after `//`

timstclair commented on container/rkt/handler.go:
```diff
@@ -250,6 +250,10 @@ func (handler *rktContainerHandler) GetStats() (*info.ContainerStats, error) {
 }
 
+func (self *rktContainerHandler) GetContainerIPAddress() string {
+	return ""
```
@yifan-gu How should this be handled?

timstclair commented on container/raw/handler.go:
```diff
@@ -256,6 +256,10 @@ func (self *rawContainerHandler) GetContainerLabels() map[string]string {
 }
 
+func (self *rawContainerHandler) GetContainerIPAddress() string {
+	return ""
```
Should this default to the host IP, or are saying that `""` implies host IP?

timstclair commented:
Can you move the collector changes to a separate PR? We try to limit PRs to a single conceptual change.

mwringe commented:
@timstclair Can you please be more specific? This PR is does a few things, all of them are improvements for the collectors. Do you just mean the second commit I applied to this issue? Or do you want the first commit also broken down more?

How receptive is cAdvisor to fixes and updates? I am wondering how difficult its going to be to get custom metrics functioning properly. Its been a bit frustrating to have pr succumb to bit rot.

mwringe commented:
@timstclair Ok, I have updated my PR again, this time all the collector stuff has been removed. Can you please take a look?

Note that I also have another 2-3 PRs to submit which are all dependent on each other. I am really hoping this is not going to take over a month each to get reviewed and merged

timstclair commented:
LGTM, but needs a rebase. Sorry for the slow turn around. Feel free to ping me directly on slack or hangouts if I take more than a day to respond on a PR.

timstclair commented:
My apologies. I missed this comment before. 

> How receptive is cAdvisor to fixes and updates?

Very. We're happy to have community contributions. Unfortunately we're a bit short-staffed on reviewers right now. I'd like to empower more community (non-Googler) maintainers, but this is slightly complicated by the current lack of a well defined scope & goal for cAdvisor. We're working on addressing this.

> I am wondering how difficult its going to be to get custom metrics functioning properly.

Hopefully it will be easier going from here. Optionally, you could have a coworker take a first pass so I can accelerate my review (especially if they're a Kubernetes maintainer). I'll look into adding more k8s maintainers to cAdvisor as well.

> Its been a bit frustrating to have pr succumb to bit rot.

:( Staying on top of all the github PRs is a challenge. Again, please feel free to ping me directly if you don't get a response in a timely manner.

Thanks for your contributions!

mwringe commented:
Rebased to master
